File: Alejandro/web/blueprints/terminal.py
from flask import Blueprint, render_template, request, jsonify
from Alejandro.web.session import get_or_create_session, Session
from Alejandro.Core.Screen import Screen, screen_type
from Alejandro.Core.Control import Control
from Alejandro.web.terminal import Terminal
from Alejandro.web.events import TerminalScreenEvent, push_event
import time

bp = Blueprint('terminal', __name__)

@screen_type
class TerminalScreen(Screen):
	"""Terminal emulator screen"""
	def __init__(self, session: 'Session'):
		# Get terminal info
		terminal_names = list(session.terminals.keys())
		current_terminal = terminal_names[session.current_terminal_index] if terminal_names and session.terminals else "none"
		
		super().__init__(
			session=session,
			title=f"Terminal - {current_terminal}",
			controls=[
				Control(
					id="new",
					text="New Terminal",
					keyphrases=["new terminal", "create terminal"],
					action=lambda s=self: s._create_new_terminal()
				),
				Control(
					id="next",
					text="Next Terminal",
					keyphrases=["next terminal"],
					action=lambda s=self: s._next_terminal()
				),
				Control(
					id="prev",
					text="Previous Terminal",
					keyphrases=["previous terminal", "prev terminal"],
					action=lambda s=self: s._prev_terminal()
				),
				session.make_back_control()
			]
		)
	
	def _create_new_terminal(self) -> None:
		"""Create a new terminal"""
		session = self.session
		terminal_id = f"terminal_{len(session.terminals) + 1}"
		# Create the new terminal
		new_terminal = Terminal(terminal_id, session.id)
		session.terminals[terminal_id] = new_terminal
		session.current_terminal_index = len(session.terminals) - 1
		self.title = f"Terminal - {terminal_id}"
		
		# Give the terminal a moment to initialize
		time.sleep(0.2)
		
		# Send a terminal switch event and replay the buffer
		new_terminal.replay_buffer()
	
	def _next_terminal(self) -> None:
		"""Switch to next terminal"""
		session = self.session
		terminal_names = list(session.terminals.keys())
		if terminal_names:
			session.current_terminal_index = (session.current_terminal_index + 1) % len(terminal_names)
			current_terminal = terminal_names[session.current_terminal_index]
			self.title = f"Terminal - {current_terminal}"
			
			# Get the terminal instance and replay its buffer
			terminal = session.terminals[current_terminal]
			# Replay the full buffer to restore all previous output
			terminal.replay_buffer()
	
	def _prev_terminal(self) -> None:
		"""Switch to previous terminal"""
		session = self.session
		terminal_names = list(session.terminals.keys())
		if terminal_names:
			session.current_terminal_index = (session.current_terminal_index - 1) % len(terminal_names)
			current_terminal = terminal_names[session.current_terminal_index]
			self.title = f"Terminal - {current_terminal}"
			
			# Get the terminal instance and replay its buffer
			terminal = session.terminals[current_terminal]
			# Replay the full buffer to restore all previous output
			terminal.replay_buffer()
			
	
	def get_template_data(self) -> dict:
		"""Get template data for rendering"""
		session = self.session
		terminal_names = list(session.terminals.keys())
		current_terminal = terminal_names[session.current_terminal_index] if terminal_names else None
		
		return {
			"terminal_id": current_terminal
		}

@bp.route(f'/{TerminalScreen.url()}')
def terminal() -> str:
	"""Terminal screen route"""
	session_id = request.args.get('session')
	terminal_id = request.args.get('terminal_id')
	
	session = get_or_create_session(session_id)
	
	# Check if we already have a terminal screen in the stack
	current_screen = session.current_or_get(TerminalScreen)
	if isinstance(current_screen, TerminalScreen):
		screen = current_screen
	else:
		# Create terminal screen and push it to the stack
		screen = TerminalScreen(session)
		session.app.screen_stack.push(screen)
	
	# Make sure terminal exists
	if not session.terminals:
		print("Creating terminal for session in route handler")
		session.terminals["main"] = Terminal("main", session.id)
		session.current_terminal_index = 0
	
	# If terminal_id is specified, switch to it
	if terminal_id and terminal_id in session.terminals:
		terminal_names = list(session.terminals.keys())
		session.current_terminal_index = terminal_names.index(terminal_id)
		screen.title = f"Terminal - {terminal_id}"
		
		# Get the terminal instance
		terminal = session.terminals[terminal_id]
		# Replay the full terminal buffer
		terminal.replay_buffer()
	
	# Get template data
	template_data = screen.get_template_data()
	print(f"Rendering terminal with data: {template_data}")
	
	return render_template(
		'terminal.html',
		screen=screen,
		session_id=session.id,
		body_class="terminal-page",
		**template_data
	)

@bp.route('/terminal/input', methods=['POST'])
def terminal_input():
	"""Handle terminal input"""
	data = request.get_json()
	session_id = data.get('session_id')
	terminal_id = data.get('terminal_id')
	input_text = data.get('input')
	
	session = get_or_create_session(session_id)
	
	# Create terminal if it doesn't exist
	if not session.terminals:
		print(f"Creating terminal '{terminal_id}' for session during input")
		session.terminals[terminal_id] = Terminal(terminal_id, session.id)
		session.current_terminal_index = 0
	
	if terminal_id in session.terminals:
		session.terminals[terminal_id].send_input(input_text)
		return jsonify({"status": "ok"})
	else:
		# Create the specific terminal if it doesn't exist
		print(f"Creating requested terminal '{terminal_id}'")
		session.terminals[terminal_id] = Terminal(terminal_id, session.id)
		session.terminals[terminal_id].send_input(input_text)
		return jsonify({"status": "ok", "created": True})

@bp.route('/terminal/resize', methods=['POST'])
def terminal_resize():
	"""Handle terminal resize"""
	data = request.get_json()
	session_id = data.get('session_id')
	terminal_id = data.get('terminal_id')
	cols = data.get('cols')
	rows = data.get('rows')
	
	session = get_or_create_session(session_id)
	
	if terminal_id in session.terminals:
		session.terminals[terminal_id].resize(cols, rows)
		return jsonify({"status": "ok"})
	else:
		return jsonify({"error": "Terminal not found"}), 404

File: Alejandro/web/static/js/terminal.js
// Initialize xterm.js
let term;
let fitAddon;
let serializeAddon;
let terminalReady = false;
let pendingData = [];

// We're not going to adjust terminal size in JS anymore

// Initialize terminal
function initTerminal() {
    console.log("Initializing xterm.js terminal");
    
    // If terminal is already initialized, just return
    if (terminalReady && term) {
        console.log("Terminal already initialized, skipping initialization");
        return;
    }
    
    // No size adjustment in JS
    
    // Check if required libraries are loaded
    if (typeof Terminal === 'undefined') {
        console.error("xterm.js Terminal is not defined! Make sure the library is loaded.");
        document.getElementById('terminal-container').innerHTML = 
            '<div style="color: red; padding: 20px;">Error: xterm.js library not loaded properly.</div>';
        return;
    }
    
    // Create terminal instance
    term = new Terminal({
        cursorBlink: true,
        theme: {
            background: '#000000',
            foreground: '#ffffff'
        },
        fontFamily: 'Menlo, Monaco, "Courier New", monospace',
        fontSize: 14,
        lineHeight: 1.2,
        scrollback: 10000,  // Significantly increased scrollback
        allowTransparency: false,
        convertEol: true,   // Convert line feed characters to carriage return + line feed
        disableStdin: false, // Ensure input is enabled
        rows: 30            // Set a specific number of rows to ensure height calculation is correct
    });

    console.log("Terminal instance created");

    // Add addons if available
    try {
        if (window.FitAddon) {
            fitAddon = new window.FitAddon.FitAddon();
            term.loadAddon(fitAddon);
            console.log("Fit addon loaded");
        } else {
            console.warn("FitAddon not available");
        }
        
        if (window.WebLinksAddon) {
            term.loadAddon(new window.WebLinksAddon.WebLinksAddon());
            console.log("WebLinks addon loaded");
        } else {
            console.warn("WebLinksAddon not available");
        }
        
        if (window.SerializeAddon) {
            serializeAddon = new window.SerializeAddon.SerializeAddon();
            term.loadAddon(serializeAddon);
            console.log("Serialize addon loaded");
        } else {
            console.warn("SerializeAddon not available");
        }
    } catch (e) {
        console.error("Error loading addons:", e);
    }

    // Get terminal container
    const container = document.getElementById('terminal-container');
    if (!container) {
        console.error("Terminal container not found!");
        return;
    }

    // Open terminal
    try {
        term.open(container);
        console.log("Terminal opened in container");
        
        if (fitAddon) {
            // Delay the fit operation slightly to ensure the container is fully rendered
            setTimeout(() => {
                fitAddon.fit();
                console.log("Terminal fitted to container");
                
                // After fitting, send the new dimensions to the server
                const dims = term.options;
                sendTerminalResize(dims.cols, dims.rows);
            }, 50);
        }
    } catch (e) {
        console.error("Error opening terminal:", e);
        container.innerHTML = 
            `<div style="color: red; padding: 20px;">Error opening terminal: ${e.message}</div>`;
        return;
    }

    // Handle terminal input
    term.onData(data => {
        sendTerminalInput(data);
    });

    // Handle window resize
    window.addEventListener('resize', () => {
        if (fitAddon && term) {
            // Debounce resize to prevent too many events
            clearTimeout(window.resizeTimer);
            window.resizeTimer = setTimeout(() => {
                console.log("Window resized, fitting terminal");
                
                // No size adjustment in JS
                
                // Then fit the terminal to its container
                fitAddon.fit();
                
                // Send terminal size to server
                const dimensions = term.options;
                sendTerminalResize(dimensions.cols, dimensions.rows);
                
                // Focus the terminal after resize
                term.focus();
            }, 100);
        }
    });

    // Mark terminal as ready
    terminalReady = true;
    console.log("Terminal ready");
    
    // Add scrollToBottom method if it doesn't exist
    if (!term.scrollToBottom) {
        term.scrollToBottom = function() {
            if (term._core && term._core.viewport) {
                term._core.viewport.scrollBarWidth = 15; // Make sure scrollbar is visible
                term._core.viewport.syncScrollArea();
                term._core.viewport.element.scrollTop = term._core.viewport.element.scrollHeight;
                console.log("Manually scrolled terminal to bottom");
            }
        };
    }

    // We will not restore buffer here on initial load
// This will be handled by the server sending initial state
// Process any pending data
if (pendingData.length > 0) {
    console.log(`Processing ${pendingData.length} pending data items`);
    pendingData.forEach(data => {
        term.write(data);
    });
    pendingData = [];
}

    // Send terminal size to server
    if (fitAddon) {
        const dimensions = term.options;
        sendTerminalResize(dimensions.cols, dimensions.rows);
    }
    
    // No need to send initial input
    
    // Terminal is ready
}

// Send terminal input to server
function sendTerminalInput(data) {
    fetch(`/terminal/input`, {
        method: 'POST',
        headers: {
            'Content-Type': 'application/json'
        },
        body: JSON.stringify({
            session_id: window.sessionId,
            terminal_id: window.terminalId,
            input: data
        })
    })
    .then(response => response.json())
    .then(data => {
        console.log('Terminal input response:', data);
    })
    .catch(error => {
        console.error('Error sending terminal input:', error);
    });
}

// Send terminal resize to server
function sendTerminalResize(cols, rows) {
    fetch(`/terminal/resize`, {
        method: 'POST',
        headers: {
            'Content-Type': 'application/json'
        },
        body: JSON.stringify({
            session_id: window.sessionId,
            terminal_id: window.terminalId,
            cols: cols,
            rows: rows
        })
    })
    .then(response => response.json())
    .then(data => {
        console.log('Terminal resize response:', data);
    })
    .catch(error => {
        console.error('Error sending terminal resize:', error);
    });
}

// Clear terminal
function clearTerminal() {
    if (term) {
        term.clear();
        // Also send clear command to server
        sendTerminalInput('\x1b[H\x1b[2J');
    }
}

// We no longer manually save terminal buffers
// Terminal state is managed by the server

// Terminal switch is now handled through the TerminalScreen controls
// No need for custom terminal switching UI

// Handle terminal data from server
function handleTerminalData(data) {
    if (data.terminal_id !== window.terminalId) {
        return; // Skip updates for other terminals
    }
    
    if (!terminalReady) {
        // Queue data until terminal is ready
        pendingData.push(data.raw_text);
        return;
    }
    
    // Write new content directly to terminal
    // Make sure we're getting proper data
    if (data.raw_text && data.raw_text.length > 0) {
        console.log(`Writing ${data.raw_text.length} chars to terminal`);
        // The terminal handles ANSI escape sequences properly
        term.write(data.raw_text);
        
        // Scroll to bottom after writing content
        setTimeout(() => {
            // Try to scroll to bottom to ensure we see the prompt
            if (term && term._core) {
                term.scrollToBottom();
            }
        }, 10);
    }
}

// Store terminal session info when navigating away
window.addEventListener('beforeunload', function(event) {
    // Only store information if we're on the terminal page
    if (window.location.pathname.includes('/terminal')) {
        console.log("Storing terminal session info before navigation");
        localStorage.setItem('terminalActive', 'true');
        localStorage.setItem('lastTerminalId', window.terminalId);
        // We no longer save terminal buffer, as it will be managed by the server
    }
});

// We don't need to save terminal state on navigation clicks anymore
// State is managed on the server side

// Custom event handler for terminal events
function handleTerminalEvents(event) {
    if (!event.data) {
        return; // Skip keepalive
    }
    
    try {
        const data = JSON.parse(event.data);
        console.log('Received event type:', data.type);
        
        if (data.type === 'TerminalScreenEvent') {
            console.log('Terminal event received for:', data.terminal_id);
            
            // Check if we need to switch terminals
            if (data.terminal_id !== window.terminalId) {
                // We no longer save terminal buffers before switching
                // Terminal state is managed by the server
                
                // Update terminal ID
                window.terminalId = data.terminal_id;
                console.log('Switching to terminal:', window.terminalId);
                
                // Update URL without reloading
                const url = new URL(window.location.href);
                url.searchParams.set('terminal_id', window.terminalId);
                window.history.replaceState({}, '', url);
                
                // Update page title
                document.title = `Terminal - ${window.terminalId} - Alejandro`;
                // Update h1 title if it exists
                const titleElement = document.querySelector('h1');
                if (titleElement) {
                    titleElement.textContent = `Terminal - ${window.terminalId}`;
                }
                
                // Clear the terminal - we'll get a full buffer from the server
                if (term) {
                    console.log(`Clearing terminal for switch to: ${window.terminalId}`);
                    term.clear();
                }
                
                console.log(`Switching to terminal: ${window.terminalId}`);
                
                // The server will send the complete buffer
                // No need to request anything else
                term.focus();
            }
            
            // Process the terminal data if there is any
            if (data.raw_text) {
                handleTerminalData(data);
            }
        }
    } catch (e) {
        console.error('Error processing terminal event:', e, event.data);
    }
}

// Initialize and set up event handling
window.addEventListener('load', function() {
    console.log("Terminal page load event fired");
    
    // Check if we're returning to the terminal page
    const wasTerminalActive = localStorage.getItem('terminalActive') === 'true';
    const lastTerminalId = localStorage.getItem('lastTerminalId');
    
    if (wasTerminalActive && lastTerminalId && window.terminalId !== lastTerminalId) {
        console.log(`Restoring previous terminal: ${lastTerminalId}`);
        window.terminalId = lastTerminalId;
        
        // Update URL to match the restored terminal
        const url = new URL(window.location.href);
        url.searchParams.set('terminal_id', lastTerminalId);
        window.history.replaceState({}, '', url);
    }
    
    // Initialize terminal
    initTerminal();
    
    // We no longer periodically save terminal buffers
    // Terminal state is managed by the server
    
    // Make sure we have the terminal ID
    console.log('Terminal page loaded with terminal ID:', window.terminalId);
    
    // Set up event handling
    if (typeof eventSource !== 'undefined') {
        console.log("Setting up terminal event listener");
        
        // Remove any existing event listener
        eventSource.removeEventListener('message', handleTerminalEvents);
        
        // Add our terminal event listener
        eventSource.addEventListener('message', handleTerminalEvents);
    } else {
        console.error("eventSource is not defined! Make sure core.js is loaded first.");
    }
});

// Add a debug function to check terminal state
function debugTerminal() {
    console.log("Terminal debug info:");
    console.log("- terminalReady:", terminalReady);
    console.log("- term initialized:", typeof term !== 'undefined');
    console.log("- fitAddon initialized:", typeof fitAddon !== 'undefined');
    console.log("- pendingData count:", pendingData.length);
    console.log("- terminal container:", document.getElementById('terminal-container'));
    
    if (typeof term !== 'undefined') {
        console.log("- term options:", term.options);
    }
}

// Call debug after a delay
setTimeout(debugTerminal, 2000);
